stellargraph/connector/neo4j/sampler.py

```diff
@@ -0,0 +1,262 @@
+                    "     RETURN out_samples_list,",
+                    "     (CASE collect(id(out_samples)) WHEN [] THEN [ _ in range(0, {num_out_nodes} - 1) | null]",
+                    "     ELSE collect(id(out_samples)) END) AS out_samples_id_list',",
+                    "    {current_node: cur_node, num_out_nodes: {out_nodes} }",
+                    "    ) YIELD value",
+                    "RETURN apoc.coll.flatten(collect(value.out_samples_list)) AS samples_list,",
+                    '       apoc.coll.flatten(collect(value.out_samples_id_list)) AS samples_id_list",',
+                    f" {{cur_node_list: cur_node_list, in_nodes: {in_num}, out_nodes: {out_num} }}) YIELD value",
+                    "WITH apoc.coll.flatten(collect([value.samples_list])) AS cur_depth_list, subgraph + apoc.coll.flatten(collect([value.samples_id_list])) AS subgraph",
+                )
+            )
+
+            query = "\n".join((query, next_hop_query))
+
+        return "\n".join((query, "return subgraph"))
+
+    def run(
+        self, neo4j_graphdb, nodes=None, n=1, in_size=None, out_size=None, seed=None
+    ):
+        """
+        Send queries to Neo4J databases and collect sampled breadth-first walks starting from the root nodes.
+
+        Args:
+            neo4j_graphdb: (py2neo.Graph) the Neo4J Graph Database object
+            nodes:  (list) A list of root node ids such that from each node n BFWs will be generated up to the
+            given depth d.
+            n: (int) Number of walks per node id.
+            in_size: (list) The number of in-directed nodes to sample with replacement at each depth of the walk.
+            out_size: (list) The number of out-directed nodes to sample with replacement at each depth of the walk.
+            seed: (int) Random number generator seed; default is None
+
+
+        Returns:
+            A list of multi-hop neighbourhood samples. Each sample expresses multiple undirected walks, but the in-node
+            neighbours and out-node neighbours are sampled separately. Each sample has the format:
+
+                [[node]
+                 [in_1...in_n]  [out_1...out_m]
+                 [in_1.in_1...in_n.in_p] [in_1.out_1...in_n.out_q]
+                    [out_1.in_1...out_m.in_p] [out_1.out_1...out_m.out_q]
+                 [in_1.in_1.in_1...in_n.in_p.in_r] [in_1.in_1.out_1...in_n.in_p.out_s] ...
+                 ...]
+
+            where a single, undirected walk might be, for example:
+
+                [node out_i  out_i.in_j  out_i.in_j.in_k ...]
+        """
+        self._check_neighbourhood_sizes(in_size, out_size)
+        self._check_common_parameters(nodes, n, len(in_size), seed)
+
+        bfsQuery = self.directed_breadth_first_search_query_builder(in_size, out_size)
+
+        result_records = neo4j_graphdb.run(bfsQuery, {"head_nodes": nodes})
+        samples = pd.DataFrame(result_records)[0][0]
+
+        return [samples]
```

Comment:
Why does this only return a single sample? What happens if `n > 1`?

Reply:
My initial intention is to let Neo4J return all the lists of ids in a single result. 
It does not currently have sampling multiple times from a single node, but I will add this (to allow for ```n > 1``` in the next commit).